OsmAnd/src/net/osmand/plus/mapmarkers/MapMarkersDbHelper.java

```diff
@@ -0,0 +1,511 @@
+	private static final String MARKERS_COL_GROUP_KEY = "group_key";
+	private static final String MARKERS_COL_COLOR = "marker_color";
+	private static final String MARKERS_COL_NEXT_KEY = "marker_next_key";
+
+	private static final String MARKERS_TABLE_CREATE = "CREATE TABLE IF NOT EXISTS " +
+			MARKERS_TABLE_NAME + " (" +
+			MARKERS_COL_ID + " TEXT PRIMARY KEY, " +
+			MARKERS_COL_LAT + " double, " +
+			MARKERS_COL_LON + " double, " +
+			MARKERS_COL_DESCRIPTION + " TEXT, " +
+			MARKERS_COL_ACTIVE + " int, " + // 1 = true, 0 = false
+			MARKERS_COL_ADDED + " long, " +
+			MARKERS_COL_VISITED + " long, " +
+			MARKERS_COL_GROUP_NAME + " TEXT, " +
+			MARKERS_COL_GROUP_KEY + " TEXT, " +
+			MARKERS_COL_COLOR + " int, " +
+			MARKERS_COL_NEXT_KEY + " TEXT);";
+
+	private static final String MARKERS_TABLE_SELECT = "SELECT " +
+			MARKERS_COL_ID + ", " +
+			MARKERS_COL_LAT + ", " +
+			MARKERS_COL_LON + ", " +
+			MARKERS_COL_DESCRIPTION + ", " +
+			MARKERS_COL_ACTIVE + ", " +
+			MARKERS_COL_ADDED + ", " +
+			MARKERS_COL_VISITED + ", " +
+			MARKERS_COL_GROUP_NAME + ", " +
+			MARKERS_COL_GROUP_KEY + ", " +
+			MARKERS_COL_COLOR + ", " +
+			MARKERS_COL_NEXT_KEY +
+			" FROM " + MARKERS_TABLE_NAME;
+
+	private static final String GROUPS_TABLE_NAME = "map_markers_groups";
+	private static final String GROUPS_COL_ID = "group_id";
+	private static final String GROUPS_COL_NAME = "group_name";
+	private static final String GROUPS_COL_TYPE = "group_type";
+
+	private static final String GROUPS_TABLE_CREATE = "CREATE TABLE IF NOT EXISTS " +
+			GROUPS_TABLE_NAME + " (" +
+			GROUPS_COL_ID + " TEXT PRIMARY KEY, " +
+			GROUPS_COL_NAME + " TEXT, " +
+			GROUPS_COL_TYPE + " int);";
+
+	private static final String GROUPS_TABLE_SELECT = "SELECT " +
+			GROUPS_COL_ID + ", " +
+			GROUPS_COL_NAME + ", " +
+			GROUPS_COL_TYPE +
+			" FROM " + GROUPS_TABLE_NAME;
+
+	public static final String TAIL_NEXT_VALUE = "tail_next";
+	public static final String HISTORY_NEXT_VALUE = "history_next";
+
+	private final OsmandApplication context;
+
+	public MapMarkersDbHelper(OsmandApplication context) {
+		this.context = context;
+	}
+
+	private SQLiteConnection openConnection(boolean readonly) {
+		SQLiteConnection conn = context.getSQLiteAPI().getOrCreateDatabase(DB_NAME, readonly);
+		int version = conn.getVersion();
+		if (version == 0 || DB_VERSION != version) {
+			if (readonly) {
+				conn.close();
+				conn = context.getSQLiteAPI().getOrCreateDatabase(DB_NAME, false);
+			}
+			version = conn.getVersion();
+			conn.setVersion(DB_VERSION);
+			if (version == 0) {
+				onCreate(conn);
+			} else {
+				onUpgrade(conn, version, DB_VERSION);
+			}
+		}
+		return conn;
+	}
+
+	private void onCreate(SQLiteConnection db) {
+		db.execSQL(MARKERS_TABLE_CREATE);
+		db.execSQL(GROUPS_TABLE_CREATE);
+		saveExistingMarkersToDb();
+	}
+
+	private void onUpgrade(SQLiteConnection db, int oldVersion, int newVersion) {
```

Comment:
So after upgrade we drop all existing markers?

Reply:
This is an option for the development time.